File: apps/ios/Anky/Core/Mirror/AnkyPostSigner.swift
import BigInt
import CryptoKit
import CryptoSwift
import Foundation

#if SWIFT_PACKAGE
import AnkyProtocol
#endif

struct SignedAnkyPost {
    let identityVersion: String
    let accountId: String
    let signatureType: String
    let signature: String
    let requestTime: String
    let client: String
    let bodySha256: String
    let eip712Digest: String
}

enum AnkyPostSigner {
    static let domainType = "EIP712Domain(string name,string version,uint256 chainId)"
    static let requestType = "AnkyMirrorRequest(string identityVersion,address account,string method,string path,bytes32 bodyHash,uint64 requestTime,string client)"

    static func sign(
        body: Data,
        identity: WriterIdentity,
        requestTime: UInt64 = Self.nowMs(),
        client: String = "ios"
    ) throws -> SignedAnkyPost {
        let bodySha256 = AnkyHasher.sha256Hex(body).addHexPrefix()
        let digest = try eip712Digest(
            identity: identity,
            bodySha256: bodySha256,
            requestTime: requestTime,
            client: client
        )
        return SignedAnkyPost(
            identityVersion: WriterIdentity.identityVersion,
            accountId: identity.accountId,
            signatureType: WriterIdentity.signingScheme,
            signature: try identity.signDigest(digest),
            requestTime: String(requestTime),
            client: client,
            bodySha256: bodySha256,
            eip712Digest: digest.toHexString().addHexPrefix()
        )
    }

    static func eip712Digest(
        identity: WriterIdentity,
        bodySha256: String,
        requestTime: UInt64,
        client: String
    ) throws -> Data {
        let domainSeparator = keccak(
            keccak(Data(domainType.utf8)) +
            encodeString("Anky") +
            encodeString("1") +
            encodeUInt(identity.chainId)
        )

        let requestHash = try keccak(
            keccak(Data(requestType.utf8)) +
            encodeString(WriterIdentity.identityVersion) +
            encodeAddress(identity.address) +
            encodeString("POST") +
            encodeString("/anky") +
            encodeBytes32(bodySha256) +
            encodeUInt(requestTime) +
            encodeString(client)
        )

        return keccak(Data([0x19, 0x01]) + domainSeparator + requestHash)
    }

    private static func nowMs() -> UInt64 {
        UInt64(Date().timeIntervalSince1970 * 1000)
    }

    private static func keccak(_ data: Data) -> Data {
        data.sha3(.keccak256)
    }

    private static func encodeString(_ value: String) -> Data {
        keccak(Data(value.utf8))
    }

    private static func encodeAddress(_ value: String) throws -> Data {
        guard let address = Data.fromHex(value), address.count == 20 else {
            throw AnkyPostSignerError.invalidAddress
        }
        return pad32(address)
    }

    private static func encodeBytes32(_ value: String) throws -> Data {
        guard let bytes = Data.fromHex(value), bytes.count == 32 else {
            throw AnkyPostSignerError.invalidBodyHash
        }
        return bytes
    }

    private static func encodeUInt(_ value: UInt64) -> Data {
        pad32(BigUInt(value).serialize())
    }

    private static func pad32(_ data: Data) -> Data {
        guard data.count < 32 else {
            return data
        }
        return Data(repeating: 0, count: 32 - data.count) + data
    }
}

enum AnkyPostSignerError: Error, Equatable {
    case invalidAddress
    case invalidBodyHash
}

File: apps/android/app/src/main/java/inc/anky/android/core/identity/AnkyPostSigner.kt
package inc.anky.android.core.identity

import inc.anky.android.core.protocol.AnkyHasher
import org.json.JSONArray
import org.json.JSONObject
import org.web3j.crypto.StructuredDataEncoder

data class SignedAnkyPost(
    val identityVersion: String,
    val accountId: String,
    val signatureType: String,
    val signature: String,
    val requestTime: String,
    val client: String,
    val bodySha256: String,
)

object AnkyPostSigner {
    fun sign(
        body: ByteArray,
        identity: WriterIdentity,
        requestTime: String = System.currentTimeMillis().toString(),
        client: String = "android",
    ): SignedAnkyPost {
        val bodySha256 = "0x${AnkyHasher.sha256Hex(body)}"
        val typedData = typedDataJson(
            identity = identity,
            bodySha256 = bodySha256,
            requestTime = requestTime,
            client = client,
        )
        val digest = StructuredDataEncoder(typedData.toString()).hashStructuredData()
        return SignedAnkyPost(
            identityVersion = WriterIdentity.IdentityVersion,
            accountId = identity.accountId,
            signatureType = "eip712",
            signature = identity.signDigest(digest),
            requestTime = requestTime,
            client = client,
            bodySha256 = bodySha256,
        )
    }

    fun typedDataJson(
        identity: WriterIdentity,
        bodySha256: String,
        requestTime: String,
        client: String,
    ): JSONObject =
        JSONObject()
            .put("types", JSONObject()
                .put("EIP712Domain", JSONArray()
                    .put(typeField("name", "string"))
                    .put(typeField("version", "string"))
                    .put(typeField("chainId", "uint256")))
                .put("AnkyMirrorRequest", JSONArray()
                    .put(typeField("identityVersion", "string"))
                    .put(typeField("account", "address"))
                    .put(typeField("method", "string"))
                    .put(typeField("path", "string"))
                    .put(typeField("bodyHash", "bytes32"))
                    .put(typeField("requestTime", "uint64"))
                    .put(typeField("client", "string"))))
            .put("primaryType", "AnkyMirrorRequest")
            .put("domain", JSONObject()
                .put("name", "Anky")
                .put("version", "1")
                .put("chainId", identity.chainId))
            .put("message", JSONObject()
                .put("identityVersion", WriterIdentity.IdentityVersion)
                .put("account", identity.address)
                .put("method", "POST")
                .put("path", "/anky")
                .put("bodyHash", bodySha256)
                .put("requestTime", requestTime.toLong())
                .put("client", client))

    private fun typeField(name: String, type: String): JSONObject =
        JSONObject().put("name", name).put("type", type)
}
